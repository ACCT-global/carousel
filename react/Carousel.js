import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Slider from 'react-slick'
import Spinner from '@vtex/styleguide/lib/Spinner'
import spinnerStyle from '@vtex/styleguide/lib/Spinner/style.css'

import Banner from './Banner'
import Arrow from './Arrow'
import Dots from './Dots'

/**
 * Carousel component. Shows a serie of banners;
 */
class Carousel extends Component {
  constructor(props) {
    super(props)

    this.state = {
      loading: true,
    }
  }

  configureSettings() {
    const {
      autoplay,
      autoplaySpeed,
      showDots,
      showArrows,
      iconsColor,
    } = this.props

    return {
      speed: 500,
      infinite: true,
      pauseOnHover: true,
      slidesToShow: 1,
      slidesToScroll: 1,
      adaptiveHeight: false,
      autoplaySpeed: autoplaySpeed * 1000,
      dots: showDots,
      arrows: showArrows,
      autoplay,
      nextArrow: <Arrow color={iconsColor} />,
      prevArrow: <Arrow color={iconsColor} />,
      appendDots: dots => <Dots color={iconsColor} dots={dots} />,
    }
  }

  componentDidMount() {
    this.setState({ loading: false })
  }

  render() {
    const {
      height,
      banner1,
      banner2,
      banner3,
      banner4,
      banner5,
      banner6,
    } = this.props
    const { loading } = this.state
    const settings = this.configureSettings()

    const banners = [banner1, banner2, banner3, banner4, banner5, banner6]

    return (
      <div className="vtex-carousel">
        {!loading && (
          <Slider {...settings}>
            {banners.map(function(banner, i) {
              if (banner && banner.image) {
                return (
                  <div key={i} style={{ maxHeight: `${height} px` }}>
                    <Banner
                      image={banner.image}
                      mobileImage={banner.mobileImage}
                      page={banner.page}
                      description={banner.description}
                      targetParams={banner.targetParams}
                    />
                  </div>
                )
              }
            })}
          </Slider>
        )}

        {loading && (
          <div className="flex justify-around pa7">
            <div className="w3">
              <Spinner style={spinnerStyle} />
            </div>
          </div>
        )}
      </div>
    )
  }
}

Carousel.schema = {
  title: 'Carousel',
  description:
    'A simple carousel component that shows a serie of banners with images and links',
  type: 'object',
  properties: {
    autoplay: {
      type: 'boolean',
      title: 'Autoplay',
      default: true,
    },
    showDots: {
      type: 'boolean',
      title: 'Show dots',
      default: true,
    },
    iconsColor: {
      type: 'string',
      title: 'Icons color',
      default: '#ffff00',
    },
    showArrows: {
      type: 'boolean',
      title: 'Show arrows',
      default: true,
    },
    height: {
      type: 'number',
      title: 'Banner max height size (px)',
      default: 400,
    },
    autoplaySpeed: {
      type: 'number',
      title: 'Autoplay speed(sec):',
      default: 5,
      enum: [4, 5, 6],
    },
    banner1: {
      type: 'object',
      title: 'Banner 1',
      properties: {
        image: {
          type: 'string',
          title: 'Banner 1 image',
        },
        mobileImage: {
          type: 'string',
          title: 'Banner 1 mobile image',
        },
        page: {
          type: 'string',
          title: 'Banner 1 link',
        },
        description: {
          type: 'string',
          title: 'Banner 1 description',
        },
        targetParams: {
          type: 'object',
          title: 'Banner 1 target params',
          properties: {
            params: {
              type: 'string',
              title: 'Params',
            },
          },
        },
      },
    },
    banner2: {
      type: 'object',
      title: 'Banner 2',
      properties: {
        image: {
          type: 'string',
          title: 'Banner 2 image',
        },
        mobileImage: {
          type: 'string',
          title: 'Banner 2 mobile image',
        },
        page: {
          type: 'string',
          title: 'Banner 2 link',
        },
        description: {
          type: 'string',
          title: 'Banner 2 description',
        },
        targetParams: {
          type: 'object',
          title: 'Banner 2 target params',
          properties: {
            params: {
              type: 'string',
              title: 'Params',
            },
          },
        },
      },
    },
    banner3: {
      type: 'object',
      title: 'Banner 3',
      properties: {
        image: {
          type: 'string',
          title: 'Banner 3 image',
        },
        mobileImage: {
          type: 'string',
          title: 'Banner 3 mobile image',
        },
        page: {
          type: 'string',
          title: 'Banner 3 link',
        },
        description: {
          type: 'string',
          title: 'Banner 3 description',
        },
        targetParams: {
          type: 'object',
          title: 'Banner 3 target params',
          properties: {
            params: {
              type: 'string',
              title: 'Params',
            },
          },
        },
      },
    },
    banner4: {
      type: 'object',
      title: 'Banner 4',
      properties: {
        image: {
          type: 'string',
          title: 'Banner 4 image',
        },
        mobileImage: {
          type: 'string',
          title: 'Banner 4 mobile image',
        },
        page: {
          type: 'string',
          title: 'Banner 4 link',
        },
        description: {
          type: 'string',
          title: 'Banner 4 description',
        },
        targetParams: {
          type: 'object',
          title: 'Banner 4 target params',
          properties: {
            params: {
              type: 'string',
              title: 'Params',
            },
          },
        },
      },
    },
    banner5: {
      type: 'object',
      title: 'Banner 5',
      properties: {
        image: {
          type: 'string',
          title: 'Banner 5 image',
        },
        mobileImage: {
          type: 'string',
          title: 'Banner 5 mobile image',
        },
        page: {
          type: 'string',
          title: 'Banner 5 link',
        },
        description: {
          type: 'string',
          title: 'Banner 5 description',
        },
        targetParams: {
          type: 'object',
          title: 'Banner 5 target params',
          properties: {
            params: {
              type: 'string',
              title: 'Params',
            },
          },
        },
      },
    },
    banner6: {
      type: 'object',
      title: 'Banner 6',
      properties: {
        image: {
          type: 'string',
          title: 'Banner 6 image',
        },
        mobileImage: {
          type: 'string',
          title: 'Banner 6 mobile image',
        },
        page: {
          type: 'string',
          title: 'Banner 6 link',
        },
        description: {
          type: 'string',
          title: 'Banner 6 description',
        },
        targetParams: {
          type: 'object',
          title: 'Banner 6 target params',
          properties: {
            params: {
              type: 'string',
              title: 'Params',
            },
          },
        },
      },
    },
  },
}

const bannerProptype = PropTypes.shape({
  image: PropTypes.string,
  mobileImage: PropTypes.string,
  page: PropTypes.string,
  description: PropTypes.string,
})

Carousel.defaultProps = {
  height: 400,
  showArrows: true,
  showDots: true,
  autoplay: true,
  autoplaySpeed: 4,
  iconsColor: '#ffff7f',
}

Carousel.propTypes = {
  /** Should change images automatically */
  autoplay: PropTypes.bool.isRequired,
  /** How long it should wait to change the banner in secs */
  autoplaySpeed: PropTypes.number,
  /** Max height size of the banners */
  height: PropTypes.number.isRequired,
  /** Should show the dots or not */
  showDots: PropTypes.bool,
  /** Should show the arrows or not */
  showArrows: PropTypes.bool,
  /** The color of the arrows and dots */
  iconsColor: PropTypes.string,
  /** Banners that will be displayed by the Carousel
   *    image - The image url of the banner
   *    page - The page that the banner will be linking to
   *    description - The description of the image
   */
  banner1: bannerProptype,
  banner2: bannerProptype,
  banner3: bannerProptype,
  banner4: bannerProptype,
  banner5: bannerProptype,
  banner6: bannerProptype,
}

export default Carousel
