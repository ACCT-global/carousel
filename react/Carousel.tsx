import './global.css'

import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Slider } from 'vtex.store-components'

import Banner, { Props as BannerProps } from './Banner'

interface Props {
  /** Should change images automatically */
  autoplay?: boolean
  /** How long it should wait to change the banner in secs */
  autoplaySpeed: number
  /** Banners that will be displayed by the Carousel */
  banners: BannerProps[]
  /** Max height size of the banners */
  height: number
  /** Set visibility of arrows */
  showArrows: boolean
  /** Set visibility of dots */
  showDots: boolean
}

/**
 * Carousel component. Shows a serie of banners.
 */
export default class Carousel extends Component<Props> {
  public static defaultProps: Props = {
    autoplay: true,
    autoplaySpeed: 5,
    banners: [],
    height: 420,
    showArrows: true,
    showDots: true,
  }

  public static propTypes = {
    /** Should change images automatically */
    autoplay: PropTypes.bool.isRequired,
    /** How long it should wait to change the banner in secs */
    autoplaySpeed: PropTypes.number.isRequired,
    /** Banners that will be displayed by the Carousel */
    banners: PropTypes.arrayOf(
      PropTypes.shape({
        /** The description of the image */
        description: PropTypes.string,
        /** The image url of the banner */
        image: PropTypes.shape({
          desktop: PropTypes.string,
          mobile: PropTypes.string
        }),
        /** The page where the image is pointing to */
        page: PropTypes.string,
        /** Params of the url */
        params: PropTypes.object,
        /** Indicates if the route is external or internal */
        typeOfRoute: PropTypes.string,
        /** The url where the image is pointing to, in case of external route */
        url: PropTypes.string,
      })
    ),
    /** Max height size of the banners */
    height: PropTypes.number.isRequired,
    /** Set visibility of arrows */
    showArrows: PropTypes.bool,
    /** Set visibility of dots */
    showDots: PropTypes.bool,
  }

  public static getSchema = (props: Props) => {
    const autoplay = props.autoplay || false

    return {
      description: 'editor.carousel.description',
      title: 'editor.carousel.title',
      type: 'object',
      properties: { // tslint:disable-line
        autoplay: {
          default: true,
          isLayout: true,
          title: 'editor.carousel.autoplay.title',
          type: 'boolean',
        },
        autoplaySpeed: autoplay
          ? {
            default: 5,
            enum: [4, 5, 6],
            isLayout: true,
            title: 'editor.carousel.autoplaySpeed.title',
            type: 'number',
            widget: {
              'ui:options': {
                inline: true,
              },
              'ui:widget': 'radio',
            },
          }
          : {},
        banners: {
          minItems: 1,
          title: 'editor.carousel.banners.title',
          type: 'array',
          // tslint:disable-next-line:object-literal-sort-keys
          items: {
            title: 'editor.carousel.banners.title',
            type: 'object',
            // tslint:disable-next-line:object-literal-sort-keys
            properties: {
              image: {
                elements: {
                  description: {
                    default: 'oi oi oi',
                    title: 'editor.carousel.banner.image.description.title',
                  },
                  hasLink: {
                    default: 'false',
                    title: 'editor.carousel.banner.image.hasLink.title'
                  },
                  // tslint:disable-next-line:object-literal-sort-keys
                  externalRoute: {
                    title: 'editor.carousel.banner.image.externalRoute.title'
                  },
                  url: {
                    title: 'editor.carousel.banner.image.url.title'
                  },
                  page: {
                    title: 'editor.carousel.banner.image.page.title'
                  },
                  params: {
                    description: 'editor.carousel.banner.image.params.description',
                    title: 'editor.carousel.banner.image.params.title'
                  },
                  desktop: {
                    default: 'https://cdn-images-1.medium.com/max/2000/1*kt9otqHk14BZIMNruiG0BA.png',
                    maxWidth: 10000,
                    title: 'editor.carousel.banner.image.desktop.title',
                  },
                  mobile: {
                    title: 'editor.carousel.banner.image.mobile.title'
                  },
                },
                title: 'editor.carousel.banner.image.title',
                type: 'image',
              },
              // tslint:disable-next-line:object-literal-sort-keys
              brand: {
                title: 'editor.carousel.banner.brand.title',
                type: 'brand'
              },
              category: {
                title: 'editor.carousel.banner.category.title',
                type: 'category'
              },
              collection: {
                title: 'editor.carousel.banner.collection.title',
                type: 'collection'
              },
              department: {
                title: 'editor.carousel.banner.department.title',
                type: 'department'
              }
            },
          },
        },
        height: {
          default: 420,
          enum: [420, 440],
          isLayout: true,
          title: 'editor.carousel.height.title',
          type: 'number',
        },
        showArrows: {
          default: true,
          isLayout: true,
          title: 'editor.carousel.showArrows.title',
          type: 'boolean',
        },
        showDots: {
          default: true,
          isLayout: true,
          title: 'editor.carousel.showDots.title',
          type: 'boolean',
        },
      },
    }
  }

  public render() {
    const { height, banners } = this.props
    const settings = this.configureSettings()
    console.log(banners && banners[0] && banners[0].brand)
    if (!banners.length) {
      return null
    }

    return (
      <div className="vtex-carousel">
        <Slider sliderSettings={settings} leftArrowClasses="ml3 ml5-m ml8-l ml9-xl" rightArrowClasses="mr3 mr5-m mr8-l mr9-xl">
          {banners.filter(banner => banner && banner.image && (banner.image.mobile || banner.image.desktop)).map(
            (banner, i) => (
              <div key={i} style={{ maxHeight: height }}>
                <Banner height={height} {...banner} />
              </div>
            )
          )}
        </Slider>
      </div>
    )
  }

  private configureSettings() {
    const { autoplay, autoplaySpeed, showDots, showArrows } = this.props

    return {
      adaptiveHeight: false,
      arrows: showArrows,
      autoplay,
      autoplaySpeed: autoplaySpeed * 1000,
      dots: showDots,
      infinite: true,
      pauseOnHover: true,
      slidesToScroll: 1,
      slidesToShow: 1,
      speed: 500,
    }
  }
}
