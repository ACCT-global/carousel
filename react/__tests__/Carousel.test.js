/* eslint-env jest */
import React from 'react'
import { render } from 'react-testing-library'

import Carousel from '../Carousel'

describe('Carousel component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = render(<Carousel />)
  })

  it('should be rendered', () => {
    expect(wrapper).toBeDefined()
  })

  it('should match snapshot', () => {
    expect(wrapper.container).toMatchSnapshot()
  })

  it('should render items', () => {
    expect(wrapper.container.querySelectorAll('.vtex-carousel').length).toBe(1)
    expect(
      wrapper.container.querySelectorAll('.vtex-carousel__arrow-right').length
    ).toBe(1)
    expect(
      wrapper.container.querySelectorAll('.vtex-carousel__arrow-left').length
    ).toBe(1)
    expect(
      wrapper.container.querySelectorAll('.vtex-carousel__dots').length
    ).toBe(1)
  })
})
