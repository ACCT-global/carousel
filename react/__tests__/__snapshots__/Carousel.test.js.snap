// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component should match snapshot 1`] = `
<div>
  <div
    class="vtex-carousel"
  >
    <div
      style="max-height: 440px;"
    >
      <div>
        <a>
          <div
            class="vtex-carousel__img-container"
          >
            <div
              class="vtex-carousel__img-regular"
              style="max-height: 440px;"
            >
              <img
                alt="banner"
                class="w-100"
                src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-01.png"
              />
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      style="max-height: 440px;"
    >
      <div>
        <a>
          <div
            class="vtex-carousel__img-container"
          >
            <div
              class="vtex-carousel__img-regular"
              style="max-height: 440px;"
            >
              <img
                alt="david"
                class="w-100"
                src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-02.png"
              />
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      style="max-height: 440px;"
    >
      <div>
        <a>
          <div
            class="vtex-carousel__img-container"
          >
            <div
              class="vtex-carousel__img-regular"
              style="max-height: 440px;"
            >
              <img
                alt="banner"
                class="w-100"
                src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-03.png"
              />
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
